### PR TITLE
Adding support for float32 spanner writer

### PR DESCRIPTION
**Pull request by VardhanThigle**

### CODE REVIEW

File: v2/spanner-common/src/main/java/com/google/cloud/teleport/v2/spanner/migrations/avro/AvroToValueMapper.java [hunk 1]
```diff
@@ -88,6 +88,9 @@ static Map<Type, AvroToValueFunction> getGsqlMap() {
     gsqlFunctions.put(
         Type.int64(),
         (recordValue, fieldSchema) -> Value.int64(avroFieldToLong(recordValue, fieldSchema)));
+    gsqlFunctions.put(
+        Type.float32(),
+        (recordValue, fieldSchema) -> Value.float64(avroFieldToFloat(recordValue, fieldSchema)));
```
Comment (Deep1998) [inline v2/spanner-common/src/main/java/com/google/cloud/teleport/v2/spanner/migrations/avro/AvroToValueMapper.java:93]: Should this be Value.float32(... on line 93?

File: v2/spanner-common/src/main/java/com/google/cloud/teleport/v2/spanner/migrations/avro/AvroToValueMapper.java [hunk 2]
```diff
@@ -192,6 +195,23 @@ static Long avroFieldToLong(Object recordValue, Schema fieldSchema) {
...
+    try {
+      if (recordValue == null) {
+        return null;
+      }
+      return Float.valueOf(recordValue.toString());
+    } catch (Exception e) {
+      throw new AvroTypeConvertorException(
+          "Unable to convert "
+              + fieldSchema.getType()
+              + " to double, with value: "
```
Comment (Deep1998) [inline v2/spanner-common/src/main/java/com/google/cloud/teleport/v2/spanner/migrations/avro/AvroToValueMapper.java:208]: to float*

File: v2/spanner-common/src/main/java/com/google/cloud/teleport/v2/spanner/migrations/avro/AvroToValueMapper.java (hunk 1 above)
Comment (VardhanThigle) [reply v2/spanner-common/src/main/java/com/google/cloud/teleport/v2/spanner/migrations/avro/AvroToValueMapper.java]: Thanks for catching this. Had missed it.
Fixed in current push.

File: v2/spanner-common/src/main/java/com/google/cloud/teleport/v2/spanner/migrations/avro/AvroToValueMapper.java [hunk 3]
```diff
@@ -192,6 +195,23 @@ static Long avroFieldToLong(Object recordValue, Schema fieldSchema) {
     }
   }
 
+  static Float avroFieldToFloat(Object recordValue, Schema fieldSchema) {
```
Comment (Deep1998) [inline v2/spanner-common/src/main/java/com/google/cloud/teleport/v2/spanner/migrations/avro/AvroToValueMapper.java:198]: nit: maybe rename the method to float32 and the below method to float64? Its more clear that way.

File: v2/spanner-common/src/test/java/com/google/cloud/teleport/v2/spanner/migrations/avro/AvroToValueMapperTest.java [hunk 4]
```diff
@@ -168,6 +168,44 @@ public void testAvroFieldToDouble_UnsupportedType() {
...
+    assertEquals(inputValue, result);
+
+    /* Test a valid double to float mapping */
+    result = AvroToValueMapper.avroFieldToFloat((Double) 3.14, SchemaBuilder.builder().floatType());
+    assertEquals("Test float input", (Float) 3.14f, result);
+
+    result = AvroToValueMapper.avroFieldToFloat("456.346", SchemaBuilder.builder().doubleType());
+    assertEquals("Test string input", (Float) 456.346f, result);
+
+    result = AvroToValueMapper.avroFieldToFloat(Long.MAX_VALUE, SchemaBuilder.builder().longType());
```
Comment (Deep1998) [inline v2/spanner-common/src/test/java/com/google/cloud/teleport/v2/spanner/migrations/avro/AvroToValueMapperTest.java:185]: Do we want to add some edge cases for float testing the limits of the range? If we have some for double, we should add that too, else we can maybe create a ticket to track a todo.

**Comment by Deep1998** [comment]
Overall LGTM

**Review by Deep1998 — APPROVED** [review]

File: v2/spanner-common/src/test/java/com/google/cloud/teleport/v2/spanner/migrations/avro/AvroToValueMapperTest.java (hunk 4 above)
Comment (VardhanThigle) [reply v2/spanner-common/src/test/java/com/google/cloud/teleport/v2/spanner/migrations/avro/AvroToValueMapperTest.java:185]: We dont have for double. But I can try for float.

File: v2/spanner-common/src/main/java/com/google/cloud/teleport/v2/spanner/migrations/avro/AvroToValueMapper.java (hunk 3 above)
Comment (VardhanThigle) [reply v2/spanner-common/src/main/java/com/google/cloud/teleport/v2/spanner/migrations/avro/AvroToValueMapper.java]: Fixed. Thanks . Float is usually float32 unless specified, but it follows spanner convention to call out 32.

File: v2/spanner-common/src/test/java/com/google/cloud/teleport/v2/spanner/migrations/avro/AvroToValueMapperTest.java [hunk 5]
```diff
@@ -168,6 +168,68 @@ public void testAvroFieldToDouble_UnsupportedType() {
...
+  @Test
+  public void testAvroFieldToFloat_ValidFloat32Input() {
+    Float inputValue = 5.75f;
+    Float result =
+        AvroToValueMapper.avroFieldToFloat32(inputValue, SchemaBuilder.builder().doubleType());
+    assertEquals(inputValue, result);
+
+    /* Test a valid double to float mapping */
+    result =
+        AvroToValueMapper.avroFieldToFloat32((Double) 3.14, SchemaBuilder.builder().floatType());
```
Comment (bharadwaj-aditya) [inline v2/spanner-common/src/test/java/com/google/cloud/teleport/v2/spanner/migrations/avro/AvroToValueMapperTest.java:180]: should this be Float instead of Double ?

File: v2/spanner-common/src/test/java/com/google/cloud/teleport/v2/spanner/migrations/avro/AvroToValueMapperTest.java [hunk 6]
```diff
@@ -168,6 +168,68 @@ public void testAvroFieldToDouble_UnsupportedType() {
...
+            Float.POSITIVE_INFINITY, SchemaBuilder.builder().doubleType());
+    assertEquals("Test positive infinity Float", (Float) Float.POSITIVE_INFINITY, result);
+
+    result =
+        AvroToValueMapper.avroFieldToFloat32(
+            Float.NEGATIVE_INFINITY, SchemaBuilder.builder().doubleType());
+    assertEquals("Test negetive infinity Float", (Float) Float.NEGATIVE_INFINITY, result);
+
+    result =
+        AvroToValueMapper.avroFieldToFloat32(Long.MAX_VALUE, SchemaBuilder.builder().longType());
```
Comment (bharadwaj-aditya) [inline v2/spanner-common/src/test/java/com/google/cloud/teleport/v2/spanner/migrations/avro/AvroToValueMapperTest.java:208]: long can store 64 bits. I'm not sure it fits in float 32

**Review by bharadwaj-aditya — DISMISSED** [review]
Overall looks fine. A couple of things to check in the test cases

File: v2/spanner-common/src/test/java/com/google/cloud/teleport/v2/spanner/migrations/avro/AvroToValueMapperTest.java (hunk 5 above)
Comment (VardhanThigle) [reply v2/spanner-common/src/test/java/com/google/cloud/teleport/v2/spanner/migrations/avro/AvroToValueMapperTest.java:180]: This is `double` on purpose. As we have discussed in comments, we test cross type mapping of a double input as a float output - this is to handle the case the source and destination schemas happen  to be double->float32

File: v2/spanner-common/src/test/java/com/google/cloud/teleport/v2/spanner/migrations/avro/AvroToValueMapperTest.java (hunk 6 above)
Comment (VardhanThigle) [reply v2/spanner-common/src/test/java/com/google/cloud/teleport/v2/spanner/migrations/avro/AvroToValueMapperTest.java]: Thanks. `Long.MAX_VALUE` came in as I used it as a template to generate float.max and min by search replace.
Removed the same.

**Review by bharadwaj-aditya — APPROVED** [review]
LGTM
